### ISSUE TITLE
Proposal: Replace @lint b/c of Elixir 1.4 compiler warnings

### ISSUE THREAD

**Issue by olafura** [rrrene/credo#291]
We are using warnings as errors and now Elixir 1.4 is complaining about:
warning: module attribute @lint was set but never used

I'm not sure if this is the right place to report this or if it should be reported to
the Elixir github issue tracker.

**Comment by aaronjensen** [comment]
A short-term workaround is to do `_ = @lint` in the function that has it

**Comment by rrrene** [comment]
Obviously the warning is correct, because the compiler does not know about our preference to have `@lint` as a special attribute like `@doc`. So it is not the fault of the compiler. But I am not sure that Credo can tell the compiler to treat `@lint` in a special way ...

@josevalim what's the right play here? can we influence the warnings?

**Comment by josevalim** [comment]
@rrrene if you are parsing `@lint` out, instead of executing the code, then there is no reason for it to be code. It could be a comment such as `# credo: ...` or anything you prefer. Would that work?

**Comment by josevalim** [comment]
We could also support a compiler option that does not warn on some attributes but then it would be more configuration for users of credo.

**Comment by olafura** [comment]
@josevalim A compiler option would work well for us

**Comment by rrrene** [comment]
> if you are parsing @lint out, instead of executing the code, then there is no reason for it to be code. 

We deliberately decided for `@lint` to be close to `@doc`. Neither represents code that's being executed at runtime of the program. But both control the execution and generated output of tools, namely `ex_doc`and `credo`. I like to think there is a reason why documentation is stored in module attributes rather than function-preceding comments (and I guess it is more than a technical reason).

`@doc false` is a very nice way of expressing that there is no intention of documenting a function. So we thought that an expression like `@lint false` feels natural and idiomatic to Elixir users.

> It could be a comment such as `# credo: ...` or anything you prefer. Would that work?

I would *really* prefer an outcome where there would be a way for Credo users to (explicitly) exclude warnings for unused `@lint` attributes via their config. 

> We could also support a compiler option that does not warn on some attributes but then it would be more configuration for users of credo

Very much: yes, please! We are discouraging the use of `@lint` anyway (in favor of better configuration or filing an issue here), so the added burden of configuration is OK in my book.

> we should not special case @lint out of the box.

Absolutely with you here. No special treatment needed! :+1:

**TL;DR** `# credo: false` feels wrong and non-idiomatic, just like `# :nodoc:` did in RDoc.

**Comment by josevalim** [comment]
> We deliberately decided for @lint to be close to @doc. Neither represents code that's being executed at runtime of the program. 

Let me clarity a bit. `@doc` is actually executable code. I can do this:

```elixir
string = "hello"
@doc string
```

Could I do something similar with lint?

```elixir
value = ...
@lint value
```

If I cannot, then it means `@lint` is quite different from `@doc` because it is parsed out and not executed.

In any case, it is not my responsibility to decide if you should use `@lint` or something else, so I will be glad to add a compiler option where you can whitelist some module attributes. Please open up an issue in Elixir if that's case. Unfortunately, such feature would only be available on 1.4.1, so it may take one or two weeks.

**Comment by rrrene** [comment]
> If I cannot, then it means `@lint` is quite different from `@doc` because it is parsed out and not executed.

You are right. I misrepresented `@doc` in my earlier statement. 

> I will be glad to add a compiler option where you can whitelist some module attributes. Please open up an issue in Elixir if that's case.

Perfect. I will submit an issue later today! :+1:

**Comment by aaronjensen** [comment]
Since we are discussing the merits of @lint vs #lint, I want to chime in with a couple things that I have felt pain with in the current @lint world. First caveat is I'm a heavy eslint user so I may be biased. 

1. #lint would allow you to target an individual line. Function wise is usually fine but sometimes I want finer grain control. 

2. I have not had success with using @lint to disable a rule on a module attribute. A common thing in our app is wanting to disable line length for a module attribute of, say, a long URL. 

3. Going against the compiler/requiring configuration "feels" wrong. @doc is metadata about the code. @lint isn't really, they are instructions to a tool. I know that feeling is subjective, but I don't think the categorization is.

**Comment by rrrene** [comment]
@aaronjensen @xtian @rubencaro @kyleaa @tsubery @plackemacher @navinpeiris @danielberkompas @amokan @hauleth @paulfioravanti @jclem You're probably right. 

I have to admit that I like `@lint false` better than `# credo :ignore_next_line`, BUT that I have no rational argument for that. Maybe I just wanted it to *look* like `@doc` - not a valid technical reason, but I am a part-time human after all :wink:

So let's talk about how the syntax of `# credo ...` should work. Judging from feedback on the current implementation, I would imagine we want to support disabling Credo (all checks or just some checks) for (a) an entire file or (b) just the line where the issue was reported.

WDYT? Any suggestions?

**Comment by hauleth** [comment]
@rrrene you can keep `#lint false` I think it would be good idea to allow simple update via:

    find . -name \*.ex -print0 | xargs -0 sed -i '' 's/@lint/#lint/g'

**Comment by rrrene** [comment]
@hauleth The problem is that the semantics would have to change IMHO. While `@lint` was based on the idea "do this for the next function", the comment syntax will be based on "do this for the next line" or the entire file. I am hesitant to mix the old meaning with a new syntax, just for the sake of easier updates (we are pre-1.0, after all).

**Comment by xtian** [comment]
@rrrene Have you considered a syntax similar to what Rubocop uses?

For example:
```elixir
# credo:disable
def function do
  # All lints disabled
end
# credo:enable

# credo:disable Credo.Check.Design.TagTODO 
def function do
  # TagTODO lint disabled
end
# credo:enable Credo.Check.Design.TagTODO 
```

or something similar?

**Comment by rrrene** [comment]
@xtian That seems really verbose, like it would get annoying pretty quickly. I'd much rather see people exclude *specific* and *rare* cases, where they exempt a line or a third party file from the analysis regarding a single check. If that does not suffice, they should report an issue, because it is likely that Credo can be improved to help them mute those failing checks. 

What I would want to avoid is people plastering control statements for my third party tool all over their codebases.

**Comment by xtian** [comment]
I would think that a more verbose syntax would actually discourage widespread usage of the control statements. I know it does for me with Rubocop.

**Comment by hauleth** [comment]
@rrrene and my idea is to apply such comment to next expression

```elixir
#lint false
def foo do
  # …
end
```

Is just one expression.

**Comment by aaronjensen** [comment]
I'd take cues from existing linting tools as @xtian suggests. Either eslint or rubocop. I'm personally a fan of eslint's, but that could be stockholm syndrome.

To disable the next line:

```elixir
# credo-disable-next-line
# or for a specific check
# credo-disable-next-line Credo.Check.Design.TagTODO
```

Implementing credo-disable-line would probably blow line lengths in Elixir quickly so you may be able to just use `credo-disable-line` to mean "next line". Not sure about how much cognitive dissonance that'd cause.

For the entire file/remainder of the file:

``` elixir
# credo-disable
# or for a specific check 
# credo-disable Credo.Check.Design.TagTODO
```

Some other things:

* I'd suggest avoiding `# lint`, it presumes one would only use a single lint tool. `# credo-disable` is more explicit.
* I'd suggest avoiding using things like atoms in the comments `# credo :disable`, they're not code and that may confuse people.

**Comment by jclem** [comment]
I appreciate that the verboseness of explicit disable/enable discourages using them, ideally disabling credo for an entire function is avoided altogether.

```elixir
# credo:disable
def foo do
end
# credo:enable
```

Having used eslint and rubicob both relatively extensively, I've always been okay with that. I'd suggest these flags:

```elixir
# credo:disable (optional list of disabled, default to all)
def foo do
end
# credo:enable (same as above)

def bar do
  # credo:disable-next-line
  this_line_is_disabled
  this_line_is_enabled
  this_line_is_disabled # credo:disable-line
end
```

One drawback of this is that people will write `foo # credo:disable` on accident, and potentially disable all of credo for the remainder of the module when they meant to disable only the line `foo`. An alternative is to keep `# credo:disable-next-line`, but have `# credo:disable` only disable the line it's on if preceded by non-whitespace.

**Comment by jeremyowensboggs** [comment]
Is there a link to the proposal in elixir to disable this warning?

**Comment by rrrene** [comment]
@jeremyowensboggs Nope, I never formalized that proposal. We will take the config-comment-route.

**Comment by rrrene** [comment]
Here is an update of where we are taking the config comments. And thx for all the valuable input above, I wrote a response at the time, but obviously never submitted the comment :'(

```elixir
defmodule OtherModule do
  # credo:disable-for-next-line
  defmacro fooBarCool do
    {:ok} = File.read
    # credo:disable-for-previous-line
  end

  # credo:disable-for-this-file
  @my_module_attr 42

  # credo:disable-for-lines:4
  @doc false
  def bar do
    :ok
  end

  defp baz do
    :ok
  end
  # credo:disable-for-lines:-3 Credo.Check.Readability.MaxLineLength
end
```

You can 

1. disable the following line
2. disable the previous line
3. disable Credo for an entire file
4. disable for a given amount of lines
5. where the amount can also be negative to target previous lines

and you can name a check that should be disabled as shown in the last example above.

Any further suggestions?

**Comment by aaronjensen** [comment]
Is there a specific need for `disable-for-previous-line`? It seems unnecessary. I'm not sure about `for-lines` either, I think it'd be more clear to just disable then enable again if that was needed. I don't like having to count lines.

I'd shorten the names up a bit, taking a cue from eslint:

```
# credo:disable-next-line
# credo:disable
```

**Comment by jclem** [comment]
Is `credo:disable-for-this-file` necessary if I can just put `credo:disable` at the top of the file?

**Comment by xtian** [comment]
I agree that `for-lines` is less clear than simply denoting the start and end of the section to disable.

It is also more brittle because if your intent is to disable linting for a certain section of code, any time you change the number of lines in that section you will need count the new number of lines and update the credo comment.

**Comment by rrrene** [comment]
A big "theme" in Elixir is to go for *explicitness* in code instead of *implied behaviour*. IMHO `credo:disable-for-this-file` is more explicit than the proposed `credo:disable`, which would (a) disable the current line in case of an end-of-line comment, (b) disable the current file if put at the top, (c) disable all issues in the current file starting with the line it is encountered and (d) can be voided with `credo:enable` in case one wants to disable liniting for sections of code, but not the complete remainder of the file.

I do not want to sound antagonistic here, but the above specification does not really resonate with me. And I have thought about this for some time (obviously, as you can see in this issue) .

The use-case for these magic comments is that I want to disable Credo for that just line where it throws an issue and we can't (for whatever reason) introduce an option to make the check accommodate that use-case. That is the only situation these comment *should* be used.

In discussions like these it's always best to discuss with real world usage and actual code examples. Could anyone please share sections of code where you would want to do the things you mentioned and where the proposed solution does not suffice?

Thank you for your input. It is very valuable to this project :+1:

**Comment by jclem** [comment]
I don't think you sound antagonistic at all! I love credo and glad this discussion is happening, regardless of the outcome.

I don't think there's a use case that your proposal doesn't cover, I just found `disable-for-this-file` less explicit and a little confusing. It's actually `disable-for-the-remainder-of-this-file`, but that to me doesn't seem like the natural interpretation. That said, maybe that's just me! Which is okay.

The reason I find `credo:disable` a little more explicit is that I read it as a switch. From that point within the file forward, credo is disabled until `credo:enable` is seen. That is the single use and interpretation of that comment.

I do not, as you've interpreted it above, think using `credo:disable` to disable one line is clear enough. For single-line toggles I prefer (and am used to seeing in prior art) `credo:disable-line` and `credo:disable-next-line`.

**Comment by jclem** [comment]
I should say I don't necessarily think `disable-next-line` is really necessary. It seems to exist in prior art to avoid triggering line length warnings when `disable-line` on the line in question is used. I'd prefer simply having just three comments: the strict off/on of `disable` and `enable`, and `disable-line`. Each, of course, would accept a list of checks, but default to all checks.

**Comment by rrrene** [comment]
> It's actually `disable-for-the-remainder-of-this-file`, [...].

It's not. You might want it to be, but if you re-read the discussion carefully, it is not. It disables Credo for the entire file.

**Comment by jclem** [comment]
I suppose I was misreading the example where it appeared, since other disables came before it. I didn't realize those were all meant to be interpreted independently of one another. My mistake. If you're going to stick with `disable-for-file` I definitely think your definition is the way to go.

**Comment by aaronjensen** [comment]
> which would (a) disable the current line in case of an end-of-line comment

Sorry, my post was not clear. I was only talking about shortening those two names, `credo-disable-line` was as terse as it needed to be, so I did not include it. I certainly do not propose `credo-disable` performing double duty here. It should only have one use, disable all checks or checks specified below the point the comment is located.



> In discussions like these it's always best to discuss with real world usage and actual code examples.

One place where we felt pain was with long strings in module attributes:

```elixir
defmodule Mod do
  @foo "This is a string that is longer than the line length limitation....................................."
  @foo2 "This is a string that is longer than the line length limitation....................................."
  @foo3 "This is a string that is longer than the line length limitation....................................."

  def rest_of_code do
    # ...
  end
end
```

I'd suggest:

```elixir
defmodule Mod do
  # credo:disable Credo.Check.Readability.MaxLineLength
  @foo "This is a string that is longer than the line length limitation....................................."
  @foo2 "This is a string that is longer than the line length limitation....................................."
  @foo3 "This is a string that is longer than the line length limitation....................................."
  # credo:enable Credo.Check.Readability.MaxLineLength

  def rest_of_code do
    # ...
  end
end
```

Here's another example from existing code:

```elixir
  @lint {Credo.Check.Readability.MaxLineLength, false}
  def study do
    %{
      id: "study",
      name: "Integrated program of study",
      description: "Students take college preparatory academic courses and career-themed courses in a combined interdisciplinary three- to four-year program that prepares them for a range of college and career options. The program uses instructional strategies that support student agency and real world applications.",
      levels: [
        # Goes on for 50 lines or more...
  end
```

Ideally I could write this as:

```elixir
  @lint {Credo.Check.Readability.MaxLineLength, false}
  def study do
    # credo:disable Credo.Check.Readability.MaxLineLength
    %{
      id: "study",
      name: "Integrated program of study",
      description: "Students take college preparatory academic courses and career-themed courses in a combined interdisciplinary three- to four-year program that prepares them for a range of college and career options. The program uses instructional strategies that support student agency and real world applications.",
      levels: [
        # Goes on for 50 lines or more...
      ]
    }
    # credo:enable Credo.Check.Readability.MaxLineLength
  end
```

As you can see, a line count here would be silly. With this data structure in my real file it'd be 95. I'd rather not have to specify line counts ever, as others have stated, they are quite fragile. I would not even consider them more "explicit" in any helpful way. I read code fragments from top to bottom generally, so seeing it like this works for me. I'd much prefer to having to keep count of lines in my head.

In the particular example above, what could make things better is a `credo:disable-next-statement` (or if `disable-next-line` actually worked on statements/expressions). That way I could just tag the map and not have to worry about enabling/disabling it. I'd still probably enable/disable if I had a clump of these.

> It disables Credo for the entire file.

This allows the comment to go anywhere in the file, including at the bottom. Some may consider this a good thing, but I wouldn't. I don't want to have to look at the top/middle/bottom to find the disable comment. I think that the fact that it behaves different from prior art, and that at least one person already has found it confusing and misunderstood your intention lends some evidence to reconsider.

At the end of the day, I don't think any option is going to be much better than any other option. Where I work, all elixir developers are also javascript developers. Everyone is familiar with eslint. It is a huge project and a ton of people using it. I don't think you can go wrong by copying some of its semantics for disabling/enabling comments. That said, I use it infrequently enough that I wouldn't care too much either way. I can adapt.

To be honest, it's probably the very long names of checks that I'll find most annoying about adding these comments whatever way the api works out.

**Comment by rrrene** [comment]
@jclem @aaronjensen Thanks both of you for your input.

Contrary to others linters, you should not need these magic comments. Every time they are needed, I will consider that a bug in Credo (except of course for cases where people want to be idiosyncratically inconsistent).

The reason why I might seem a bit difficult - even in the face of so much "prior art" that has already figured this out - is that Credo *tries* to look at these questions with a fresh set of eyes.

@aaronjensen Thanks for your thoughtful and long response. It cleared up most questions I had :+1:

**Comment by dideler** [comment]
FWIW, the reason I like rubocop's style of disabling and re-enabling linting is that the extra friction which makes it verbose is dirty, and that encourages me to fix the style violation quicker. In other words, the disabling of the linter is more likely to be treated as temporary by me if disabling it looks like a smell.

**Comment by zachdaniel** [comment]
Hey guys! What is the state of this? If I tell credo to disable the next line and that line is a function head, will I still get lint errors/warnings for the contents of that function?

**Comment by rrrene** [comment]
@zachdaniel This has been implemented an should work. Are you still having this problem?

**Comment by zachdaniel** [comment]
We upgraded and this has been resolved, thanks!

**Comment by lleger** [comment]
We upgraded and this worked great for us. Thanks!
